sessiontxn/isolation/repeatable_read.go

```diff
@@ -234,7 +234,11 @@ func notNeedGetLatestTSFromPD(plan plannercore.Plan, inLockOrWriteStmt bool) boo
 	return false
 }
 
-func (p *PessimisticRRTxnContextProvider) handleAfterPessimisticLockError(lockErr error) (sessiontxn.StmtErrorAction, error) {
+func (p *PessimisticRRTxnContextProvider) handleAfterPessimisticLockError(ctx context.Context, lockErr error) (sessiontxn.StmtErrorAction, error) {
+	if err := p.retryAggressiveLockingIfNeeded(ctx); err != nil {
+		return sessiontxn.ErrorAction(err)
+	}
```

Comment:
Is `txn.retryAggressiveLockingIfNeeded` necessary for the other two pathes that error would be returned? If not should we move this part to L255-L268?

Reply:
Ah yes, I missed that. It's only needed in retrying cases. And I just noticed that for single-statement-deadlock cases, we should exit aggressive locking mode, otherwise it will still keep the lock that's blocking the other transaction.